### PR TITLE
Feat: Allowing for multiple images upload

### PR DESCRIPTION
**Pull request by omer-bar**
## What type of PR is this?

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [X] Yes

There is a discussion on the matter with the team in the InvokeAI discord server if you want to contribute to it:
https://discord.com/channels/1020123559063990373/1197563655357739149
      
## Have you updated all relevant documentation?
- [ ] Yes
- [X] No

## Description
The current state of the app is you can only upload 1 image at a time, This feature PR is to allow the user to upload multiple images through the UI and the API.
This feature also allows the user to upload whole folders of images even in nested folders.

This PR is not ready for merge, there's still work to be done on it.
TODO List:
- [ ] Fix locking of the app behavior when uploading images.
- [ ] Fix events behavior between the backend and the frontend.
- [ ] Fix Toast showing in the frontend letting the user know of the image upload process progression.
- [ ] Write tests for the Feature with pytest.
- [ ]  [If necessary] Write documentation for the upload route

## Related Tickets & Documents

- Related Issue #5529

## Merge Plan
This is a draft PR for the team to test and give feedback on, once it is done I will discuss the best merge plan for this Feature with the team.

## Added/updated tests?

- [ ] Yes
- [X] No : Tests are in the TODO list as next steps of development of the feature

## [optional] Are there any post deployment tasks we need to perform?
Not at the moment.

### CODE REVIEW

**Comment by psychedelicious** [comment]
It appears you have set your invoke root dir to be in the repo, resulting in a number of extraneous files being committed to this PR (like the database), and requiring changing `.gitignore`. Please remove all these changes and use a root dir outside the repo hierarchy.

**Comment by Millu** [comment]
@omer-bar is this ready for review? I know you mentioned you were adding more tests in the Discord

**Comment by omer-bar** [comment]
> @omer-bar is this ready for review? I know you mentioned you were adding more tests in the Discord

This is 85% done, it's working but it has some kinks in it still, mainly at the frontend part, i'v written them down in the App locking thread on Discord.

File: invokeai/app/api/routers/images.py [hunk 1]
```diff
@@ -22,6 +22,101 @@
...
+)
+async def upload_images(
+    files: List[UploadFile],
+    background_tasks: BackgroundTasks,
+    request: Request,
+    response: Response,
+    image_category: ImageCategory = Query(description="The category of the images"),
+    is_intermediate: bool = Query(description="Whether this is an intermediate images"),
+    board_id: Optional[str] = Query(default=None, description="The board to add this images to, if any"),
+    session_id: Optional[str] = Query(default=None, description="The session ID associated with this upload, if any"),
```
Comment (psychedelicious) [inline invokeai/app/api/routers/images.py:43]: `session_id` is no longer used for uploads - this was needed for the canvas at some point in the past. Can omit this. We'll need to clean it up in the single upload route too.

File: invokeai/app/services/images/images_default.py [hunk 2]
```diff
@@ -24,15 +29,220 @@
...
+
+        with ThreadPoolExecutor(max_workers=num_workers) as executor:
+            futures = [executor.submit(process_and_save_image, image) for image in upload_data_list]
+            for future in as_completed(futures):
+                try:
+                    image_dto = future.result()
+                    images_DTOs.append(image_dto)
+                    pbar.update(1)  # Update progress bar
+
+                    progress_percentage = (processed_counter / total_images) * 100
```
Comment (psychedelicious) [inline invokeai/app/services/images/images_default.py:185]: We already have `processed` and `total`, do we need percentage in this payload? It's trivial to derive it clientside

File: invokeai/frontend/web/src/features/gallery/store/gallerySlice.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ export const initialGalleryState: GalleryState = {
   boardSearchText: '',
   limit: INITIAL_IMAGE_LIMIT,
   offset: 0,
+  uploadedImages: [],
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/gallery/store/gallerySlice.ts:23]: It looks like we are manipulating this array but never actually using it anywhere

File: invokeai/frontend/web/src/services/api/endpoints/images.ts [hunk 4]
```diff
@@ -595,10 +698,105 @@ export const imagesApi = api.injectEndpoints({
         }
       },
     }),
+    /**
+     * Upload a multiple images.
+     */
+    uploadMultipleImages: build.mutation<
+      ImageDTO[],
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/services/api/endpoints/images.ts:705]: Mentioned on discord - this query only returns a message now, so the cache handling in `onQueryStarted` is unnecessary (and it errors in js console)

File: invokeai/frontend/web/src/services/events/types.ts [hunk 5]
```diff
@@ -202,17 +209,31 @@ export type ClientEmitUnsubscribeQueue = {
...
+  status: 'started' | 'processing' | 'done' | 'error';
+  message?: string;
+  progress?: number;
+  processed?: number;
+  total?: number;
+  errors?: string[];
+  images_uploading?: string[];
+  images_DTOs?: ImageDTO[];
+};
+
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/services/events/types.ts:222]: Looks like there are 4 distinct payload types. Could be nice to have this be a discriminated union so we get good type hints when narrowing on `status`.

I'm not asking to use this feature in this PR, but [`fastapi-events`](https://github.com/melvinkcx/fastapi-events) supports pydantic models for its payloads. And a new feature lets you [dispatch specially formed pydantic models directly](https://github.com/melvinkcx/fastapi-events?tab=readme-ov-file#option-2---using-pydantic-model).

That functionality would make it easier to add these payloads to the OpenAPI schema and we'd get typescript types for free. Something to try in the future.

File: invokeai/app/services/images/images_common.py [hunk 6]
```diff
@@ -23,6 +45,12 @@ class ImageDTO(ImageRecord, ImageUrlsDTO):
     board_id: Optional[str] = Field(
         default=None, description="The id of the board the image belongs to, if one exists."
     )
+    metadata: Optional[MetadataField] = Field(
+        default=None, description="The metadata of the image."
+    )
+    workflow: Optional[WorkflowWithoutID] = Field(
+        default=None, description="The workflow of the image."
+    )
```
[the remark this answers is not in the document]
Comment (omer-bar) [reply invokeai/app/services/images/images_common.py:53]: It has already been discussed on Discord and will be removed in the next commits

**Comment by psychedelicious** [comment]
Cleaning up old PRs and closing this. We've got a good-enough implementation of multi-uploads released recently. Thanks for your work on this @omer-bar, sorry it didn't make it across the finish line.

**Pull request closed without merging**